beacon-chain/sync/rpc_blob_sidecars_by_root.go

```diff
@@ -0,0 +1,125 @@
+	"github.com/prysmaticlabs/prysm/v4/consensus-types/primitives"
+	"github.com/prysmaticlabs/prysm/v4/encoding/bytesutil"
+	"github.com/prysmaticlabs/prysm/v4/monitoring/tracing"
+	eth "github.com/prysmaticlabs/prysm/v4/proto/prysm/v1alpha1"
+	"github.com/prysmaticlabs/prysm/v4/time/slots"
+	"go.opencensus.io/trace"
+)
+
+func blobMinReqEpoch(finalized, current primitives.Epoch) primitives.Epoch {
+	// max(finalized_epoch, current_epoch - MIN_EPOCHS_FOR_BLOB_SIDECARS_REQUESTS, DENEB_FORK_EPOCH)
+	denebFork := params.BeaconConfig().DenebForkEpoch
+	var reqWindow primitives.Epoch
+	if current > params.BeaconNetworkConfig().MinEpochsForBlobsSidecarsRequest {
+		reqWindow = current - params.BeaconNetworkConfig().MinEpochsForBlobsSidecarsRequest
+	}
+	if finalized >= reqWindow && finalized > denebFork {
+		return finalized
+	}
+	if reqWindow >= finalized && reqWindow > denebFork {
+		return reqWindow
+	}
+	return denebFork
+}
+
+// blobSidecarByRootRPCHandler handles the /eth2/beacon_chain/req/blob_sidecars_by_root/1/ RPC request.
+// spec: https://github.com/ethereum/consensus-specs/blob/a7e45db9ac2b60a33e144444969ad3ac0aae3d4c/specs/deneb/p2p-interface.md#blobsidecarsbyroot-v1
+func (s *Service) blobSidecarByRootRPCHandler(ctx context.Context, msg interface{}, stream libp2pcore.Stream) error {
+	ctx, span := trace.StartSpan(ctx, "sync.blobSidecarByRootRPCHandler")
+	defer span.End()
+	ctx, cancel := context.WithTimeout(ctx, ttfbTimeout)
+	defer cancel()
+	SetRPCStreamDeadlines(stream)
+	log := log.WithField("handler", p2p.BlobSidecarsByRootName[1:]) // slice the leading slash off the name var
+	ref, ok := msg.(*types.BlobSidecarsByRootReq)
+	if !ok {
+		return errors.New("message is not type BlobSidecarsByRootReq")
+	}
+
+	blobIdents := *ref
+	if uint64(len(blobIdents)) > params.BeaconNetworkConfig().MaxRequestBlobsSidecars {
+		s.cfg.p2p.Peers().Scorers().BadResponsesScorer().Increment(stream.Conn().RemotePeer())
+		s.writeErrorResponseToStream(responseCodeInvalidRequest, types.ErrMaxBlobReqExceeded.Error(), stream)
+		return types.ErrMaxBlobReqExceeded
+	}
+	// Sort the identifiers so that requests for the same blob root will be adjacent, minimizing db lookups.
+	sort.Sort(blobIdents)
+
+	batchSize := flags.Get().BlobBatchLimit
+	var ticker *time.Ticker
+	if len(blobIdents) > batchSize {
+		ticker = time.NewTicker(time.Second)
+	}
+	minReqEpoch := blobMinReqEpoch(s.cfg.chain.FinalizedCheckpt().Epoch, slots.ToEpoch(s.cfg.clock.CurrentSlot()))
+
+	buff := struct {
+		root [32]byte
+		scs  []*eth.BlobSidecar
+	}{}
+	for i := range blobIdents {
+		if err := ctx.Err(); err != nil {
+			closeStream(stream, log)
+			return err
+		}
+
+		// Throttle request processing to no more than batchSize/sec.
+		if i != 0 && i%batchSize == 0 && ticker != nil {
+			<-ticker.C
+		}
+		s.rateLimiter.add(stream, 1)
+		root, idx := bytesutil.ToBytes32(blobIdents[i].BlockRoot), blobIdents[i].Index
+		if root != buff.root {
```

Comment:
maybe its better to explicitly just check that it is not a zeroed 32-byte root ?

Reply:
You mean we should check to make sure that the request doesn't contain a zero-value block root?

This code is checking to see if the block root for the sidecar id being currently processed is the same as the last step of the iteration. That way we can look up all the sidecars for the given root at once (because they are currently stored together in the db) and iterate through them - this is why the idents are sorted by root `sort.Sort(blobIdents)`.

If we do get a request for a zero root, we'll skip it as a side effect of the other invariants - the zero root will sort to the beginning of the list, so the initial value of `buff.scs` will be a nil list, which will have a length of zero, so we'll skip it here:
```
		if idx >= uint64(len(buff.scs)) {
			continue
		}
```
We could check for it and explicitly skip it. But if things did somehow change such that these invariants no longer prevented it from being skipped, then we would fail to look up the zero root from the db, and just return an error. So to me it feels like clutter to check for it.